EIP-2537 precompile for bls12 381 curve operations

lorenzogentile404 opened the pull request.

lorenzogentile404 committed:
added structure of BLS module and updated reference to constraints

lorenzogentile404 committed:
added index max and phase methods

lorenzogentile404 committed:
updated constraints

lorenzogentile404 committed:
fixed total size data usage

lorenzogentile404 committed:
added utils and specialized computation for point evaluation

lorenzogentile404 committed:
refined utilities and added computation for g1 add

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/bls/BlsOperation.java:
```diff
@@ -0,0 +1,449 @@
+
+@Accessors(fluent = true)
+public class BlsOperation extends ModuleOperation {
+  final EWord BLS_PRIME_HI = EWord.of(BigInteger.valueOf(BLS_PRIME_3), BLS_PRIME_2);
```
static ?

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/bls/BlsOperation.java:
```diff
@@ -0,0 +1,449 @@
+  final EWord POINT_EVALUATION_PRIME =
+      EWord.of(POINT_EVALUATION_PRIME_HI, POINT_EVALUATION_PRIME_LO);
+
+  private final int SIZE_SMALL_POINT = LLARGE * (CT_MAX_SMALL_POINT + 1);
```
same, could be static static ?

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/bls/BlsOperation.java:
```diff
@@ -0,0 +1,449 @@
+  private final List<Boolean> isInfinity;
+
+  // WCP interaction
+  private final List<Boolean> wcpFlag;
```
It's easier to read if you do a List<WcpInteraction> where WcpInteraction contans all the above. See for exemple in OOB

letypequividelespoubelles replied:
this comment still apply

letypequividelespoubelles replied:
commet still apply :)

lorenzogentile404 replied:
you mean implementing something similar to `OobExoCall` @letypequividelespoubelles?

letypequividelespoubelles reviewed (commented):
a

lorenzogentile404 committed:
propagated mint bit and is infinity conditions and added computation for g1 msm

lorenzogentile404 committed:
added specialized computation for g2 case

lorenzogentile404 committed:
added specialized computatation for maps

lorenzogentile404 committed:
filled g1 case result rows

lorenzogentile404 committed:
filled remaining result rows

lorenzogentile404 committed:
filled part of the trace

lorenzogentile404 committed:
added ct, ctMax, isFirstInput, isSecondInput logic

lorenzogentile404 committed:
filled more columns of the trace

lorenzogentile404 committed:
managed trivial and non trivial pairings columns

lorenzogentile404 committed:
updated constraints

lorenzogentile404 committed:
Merge branch 'arith-dev' into eip-2537-precompile-for-bls12-381-curve-operations

lorenzogentile404 committed:
updated constraints

lorenzogentile404 committed:
updated constraints

lorenzogentile404 committed:
imported constants from London

lorenzogentile404 committed:
added structure for mext computation

lorenzogentile404 committed:
Merge branch 'arith-dev' into eip-2537-precompile-for-bls12-381-curve-operations

lorenzogentile404 committed:
fixed minor bugs

lorenzogentile404 committed:
made logic for point is at infinity, point is on curve and point is in subgroup more modular

lorenzogentile404 committed:
added classes to deal with Fp and Fp2 points

lorenzogentile404 committed:
added logic for subgroup memership checks

lorenzogentile404 committed:
fixed bugs related to field operations

lorenzogentile404 committed:
fixed comments

lorenzogentile404 committed:
fixed bug related to points addition

lorenzogentile404 committed:
extracted BlsUtils

lorenzogentile404 committed:
fixed addition operations

lorenzogentile404 committed:
used generics to generalize fields and points behaviour

lorenzogentile404 committed:
fixed issues related to casting in class using java generics

lorenzogentile404 committed:
made multiplication method over elliptic curve more readable

lorenzogentile404 committed:
fixed initialization issue

lorenzogentile404 committed:
used successBit to deduce mextBit

lorenzogentile404 committed:
added todos

lorenzogentile404 committed:
removed limb list

lorenzogentile404 committed:
fixed preconditions

lorenzogentile404 committed:
filled circuit selectors and fixed mext bit computation

lorenzogentile404 committed:
updated constraints

lorenzogentile404 committed:
Merge branch 'arith-dev' into eip-2537-precompile-for-bls12-381-curve-operations

lorenzogentile404 committed:
partially implemented bls oob calls

lorenzogentile404 committed:
updated constraints and used corresponding constants

lorenzogentile404 committed:
added calls for fixed size fixed cost oob precompiles

lorenzogentile404 committed:
added msm oob calls

lorenzogentile404 committed:
added bls pairing check oob call

lorenzogentile404 committed:
used bls pairing check oob call

lorenzogentile404 committed:
updated constraints

lorenzogentile404 committed:
Merge branch 'arith-dev' into eip-2537-precompile-for-bls12-381-curve-operations

lorenzogentile404 committed:
added bls reference table

lorenzogentile404 committed:
used bls reference table and filled oob trace

lorenzogentile404 commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/oob/precompiles/common/bls/fixedSizeFixedGasCost/BlsPointEvaluationOobCall.java:
```diff
@@ -0,0 +1,56 @@
+  @Override
+  public int ctMax() {
+    return CT_MAX_BLS_MAP_FP2_TO_G2;
+  }
```
Thanks Cursor <3.

lorenzogentile404 committed:
fixed constant issue

lorenzogentile404 committed:
added mmu calls

lorenzogentile404 committed:
added call to mmu for bls precompiles

lorenzogentile404 committed:
completed mmu calls for bls

lorenzogentile404 committed:
used new constants

lorenzogentile404 committed:
Merge branch 'arith-dev' into eip-2537-precompile-for-bls12-381-curve-operations

lorenzogentile404 committed:
fixed constants usage

lorenzogentile404 committed:
fixed tracing for map fp2 to g2

lorenzogentile404 committed:
used updated constantas and updated constraints

lorenzogentile404 committed:
Merge branch 'arith-dev' into eip-2537-precompile-for-bls12-381-curve-operations

lorenzogentile404 committed:
updated constraints

lorenzogentile404 committed:
added bls precompiles to PrecompileScenarioFragment

lorenzogentile404 committed:
fixed naming and updated constraints

lorenzogentile404 committed:
Merge branch 'arith-dev' into eip-2537-precompile-for-bls12-381-curve-operations

lorenzogentile404 committed:
added headers

lorenzogentile404 committed:
used besu methods to check if points are on curve and in subgroup

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/tables/bls/BlsRt.java:
```diff
@@ -0,0 +1,352 @@
+          Map.entry(128, 524));
+
+  public static int getMsmDiscount(final int instruction, final int numInputs) {
+    if (numInputs < 1 || numInputs > 128) {
```
better use checkArgument

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/tables/bls/BlsRt.java:
```diff
@@ -0,0 +1,352 @@
+    if (numInputs < 1 || numInputs > 128) {
+      throw new IllegalArgumentException("Number of inputs must be between 1 and 128");
+    }
+    if (instruction == OOB_INST_BLS_G1_MSM) {
```
switch case with default branch is more lisible

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/tables/bls/BlsRt.java:
you need to add this to the refTable in the HUB, see how it's done for PowerRt for exemple (exist only post cancun lilke this one)

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/types/Conversions.java:
```diff
@@ -24,7 +24,7 @@
 
 public class Conversions {
-  public static final Bytes ZERO = Bytes.EMPTY;
+  public static final Bytes ZERO = Bytes.of(0);
```
what's the implication of changing this ?

letypequividelespoubelles replied:
?

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/bls/Bls.java:
```diff
@@ -0,0 +1,90 @@
+
+  @Override
+  public void commit(Trace trace) {
+    int stamp = 0;
```
do we really need the stamp ? I mean, can't it be computed by corset with a 'defcomputedcolumn' declaration ?

letypequividelespoubelles replied:
comment still apply

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/bls/BlsOperation.java:
```diff
@@ -0,0 +1,1117 @@
+      case PRC_BLS_G2_MSM -> blsOperation.handleBlsG2Msm();
+      case PRC_BLS_PAIRING_CHECK -> blsOperation.handleBlsPairingCheck();
+      case PRC_BLS_MAP_FP_TO_G1 -> blsOperation.handleBlsMapFpToG1();
+      case PRC_BLS_MAP_FP2_TO_G2 -> blsOperation.handleBlsMapFp2ToG2();
```
better to add a default branch that throws an exception

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/Hub.java:
```diff
@@ -206,6 +207,7 @@ public int spillage(Trace trace) {
   private final Add add = setAdd();
   private final Bin bin = new Bin();
   private final Blockhash blockhash = new Blockhash(this, wcp);
+  public final Bls bls = new Bls(wcp);
```
Bls appears only in Cancun, so shouldn't be before; See how it's done for RlpUtils for exemple

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/scenario/PrecompileScenarioFragment.java:
```diff
@@ -123,6 +155,18 @@ public boolean isEcdataPrecompile() {
     }
 
+    public boolean isBlsPrecompile() {
+      return this.isAnyOf(
```
isn't the same list as `public static final Set<Address> BLS_PRECOMPILES = ...` in BLS moduel ? If yes, better use this static list, better to have only one ref

lorenzogentile404 replied:
This is over `PrecompileFlag`, while the other one on `Address`.

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/tables/bls/BlsRt.java:
```diff
@@ -0,0 +1,352 @@
+          .prcName(OOB_INST_BLS_G2_MSM)
+          .numInputs(numInputs)
+          .discount(discount)
+          .validateRow();
```
theer are no IOMF column ? I guess it's because input are somewhere constrained to be !=0 ?

letypequividelespoubelles reviewed (commented):
some generic comments, didn't look at all at the specific implementation, just the architecture

lorenzogentile404 committed:
addressed review comments

lorenzogentile404 committed:
added simple bls g1 add test

lorenzogentile404 committed:
added simple bls g2 add test

DavePearce committed:
initialise BlsRT as a reference table

Previously, the BlsRT reference table was not being initialised and this
meant its columns were not included in the final trace.  This now
initialises BlsRT as for other modules.

lorenzogentile404 committed:
set BlsRt for Cancun

lorenzogentile404 committed:
spotless

lorenzogentile404 committed:
Merge branch 'arith-dev' into eip-2537-precompile-for-bls12-381-curve-operations

lorenzogentile404 committed:
removed defcomputedcolumn from tracing

lorenzogentile404 committed:
Merge branch 'arith-dev' into eip-2537-precompile-for-bls12-381-curve-operations

lorenzogentile404 committed:
made tracing of oob and hub fork dependent

lorenzogentile404 committed:
partially fixed connection between hub and bls

lorenzogentile404 committed:
used EXTCODECOPY to write input in memory

lorenzogentile404 committed:
added point evalution precompile test

lorenzogentile404 committed:
Merge branch 'arith-dev' into eip-2537-precompile-for-bls12-381-curve-operations

lorenzogentile404 committed:
added BlsPrecompilesTest

lorenzogentile404 committed:
Merge branch 'arith-dev' into eip-2537-precompile-for-bls12-381-curve-operations

lorenzogentile404 committed:
updated constraints

lorenzogentile404 committed:
fixed minor issues

lorenzogentile404 committed:
Merge branch 'arith-dev' into eip-2537-precompile-for-bls12-381-curve-operations

lorenzogentile404 committed:
updated constraints

lorenzogentile404 committed:
removed duplicated method

lorenzogentile404 committed:
partially added bls precompiles gas tests

lorenzogentile404 committed:
added bls precompiles to LowGastStipendPrecompileCallTests

lorenzogentile404 committed:
added todo

lorenzogentile404 committed:
computed precompile limits

lorenzogentile404 committed:
fixed bls limits

lorenzogentile404 committed:
added point evaluations effective call and failure call

lorenzogentile404 committed:
added bls calls

lorenzogentile404 committed:
Merge branch 'arith-dev' into eip-2537-precompile-for-bls12-381-curve-operations

lorenzogentile404 committed:
updated constraints

lorenzogentile404 committed:
Merge branch 'arith-dev' into eip-2537-precompile-for-bls12-381-curve-operations

lorenzogentile404 committed:
fixed limits computation

lorenzogentile404 committed:
fixed minor issues

lorenzogentile404 committed:
fixed bls limits computation

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/limits/precompiles/BlsG1AddEffectiveCall.java:
```diff
@@ -0,0 +1,41 @@
+  @Override
+  public void updateTally(final int numberEffectiveCall) {
+    checkArgument(
+        numberEffectiveCall <= 1, "can't add more than one effective precompile call at a time");
```
What does the 1 represent here ? Is this called with every effective call to the `BLS_G1_ADD` gnark circuit ?

lorenzogentile404 replied:
One call to the circuit per call to the precompile, yes.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/Hub.java:
```diff
@@ -396,7 +445,7 @@ public Hub(final ChainConfig chain) {
 
     refTableModules =
-        Stream.of(new BinRt(), setInstructionDecoder(), new ShfRt(), setPower())
+        Stream.of(new BinRt(), setBlsRt(), setInstructionDecoder(), new ShfRt(), setPower())
```
The BLS reference table being potentially null isn't an issue ?

lorenzogentile404 replied:
It is supposed to be used only in Prague, where it is initialised, that is the same approach we use for other modules.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/bls/Bls.java:
```diff
@@ -0,0 +1,178 @@
+
+  @Override
+  public String moduleKey() {
+    return "BLS";
```
The module is called `BLS_DATA` (like `EC_DATA`)

letypequividelespoubelles committed:
Merge branch 'arith-dev' into eip-2537-precompile-for-bls12-381-curve-operations

lorenzogentile404 committed:
fixed mergin issue

lorenzogentile404 committed:
Merge branch 'eip-2537-precompile-for-bls12-381-curve-operations' of github.com:Consensys/linea-tracer into eip-2537-precompile-for-bls12-381-curve-operations

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/bls/BlsOperation.java:
```diff
@@ -0,0 +1,1094 @@
+              aYRe3,
+              aYRe2,
+              aYRe1,
+              aYRe0);
```
What I don't understand is why you cut up the coordinates at this stage. For instance
```java
              aXIm3,
              aXIm2,
              aXIm1,
              aXIm0,
```
Represents a 64 byte string which may or may not correspond to a single 48 byte field element. Why do the splitting here when
- the hi/lo stuff shouldn't matter to the implementation
- you end up recombining stuff e.g. in `wcpGeneralizedCallToLT`

OlivierBBB replied:
Another example is `isLargePointOnCurve` which should take a single `Bytes` argument, or maybe two. You end up re-combining everything before calling `eip2537G2IsInSubGroup`. There should be no reason to do any of this splitting at this point.

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/MmuCall.java:
```diff
@@ -579,7 +631,9 @@ public static MmuCall partialCopyOfReturnDataForEcpairing(
-                leftPadTo(subsection.returnDataRange.extract(), TOTAL_SIZE_ECPAIRING_RESULT)))
+                leftPadTo(
+                    subsection.returnDataRange.extract(),
+                    TOTAL_SIZE_ECPAIRING_RESULT))) // TODO: is the padding necessary?
```
why this TODO ? Did you add a test for this ?

lorenzogentile404 replied:
It is probably an old comment, I'll check it out.

OlivierBBB reviewed (dismissed):
Very preliminary review. The ubiquitous  splitting of `Bytes` into 16 byte chunks and the many, many argument functions don't look right. There should be no need for hardly any such splitting, except maybe for variable size operations, and functions shouldn't take 12 arguments only to be recombined within the function.

lorenzogentile404 committed:
renamed bls to blsData

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/scenario/PrecompileScenarioFragment.java:
```diff
@@ -59,44 +62,75 @@ public enum PrecompileFlag {
-            PRC_ECMUL, Trace.PHASE_ECMUL_RESULT,
-            PRC_ECPAIRING, Trace.PHASE_ECPAIRING_RESULT,
-            PRC_BLAKE2F, Trace.PHASE_BLAKE_RESULT);
+            // MODEXP not supported
```
? why no more supported ?

lorenzogentile404 replied:
I believe I have not touched it. I'll check it out.

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/scenario/PrecompileScenarioFragment.java:
```diff
@@ -59,44 +62,75 @@ public enum PrecompileFlag {
+            Map.entry(PRC_ECADD, Trace.PHASE_ECADD_RESULT),
+            Map.entry(PRC_ECMUL, Trace.PHASE_ECMUL_RESULT),
+            Map.entry(PRC_ECPAIRING, Trace.PHASE_ECPAIRING_RESULT),
+            // BLAKE2f not supported
```
same question

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/scenario/PrecompileScenarioFragment.java:
```diff
@@ -176,6 +222,18 @@ public Trace.Hub trace(Trace.Hub trace) {
         .pScenarioPrcCds(precompileSubSection.callDataSize())
         .pScenarioPrcRao(precompileSubSection.returnAtOffset())
         .pScenarioPrcRac(precompileSubSection.returnAtCapacity());
+    // TODO: what is a more natural way to handle this?
```
do like for account / stack fragment

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/limits/precompiles/BlsC2MembershipCalls.java:
```diff
@@ -0,0 +1,39 @@
+
+  @Override
+  public String moduleKey() {
+    return "PRECOMPILE_BLS_PAIRING_CHECK_MILLER_LOOPS";
```
name not consistent with class name

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/oob/OobExoCall.java:
```diff
@@ -53,6 +60,10 @@ protected void trace(Trace.Oob trace) {
         .outgoingData3(arg2.slice(0, LLARGE))
         .outgoingData4(arg2.slice(LLARGE, LLARGE))
         .outgoingResLo(addFlag ? ZERO : result);
+    // TODO: what is a more natural way to handle this?
```
you could have if 
```
if (blsRtFlag) {
trace.blsRefTableFlag(true);}
```

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/bls/BlsOperation.java:
```diff
@@ -0,0 +1,1094 @@
+public class BlsOperation extends ModuleOperation {
+  public static final BigInteger BLS_PRIME =
+      Bytes.concatenate(
+              Bytes.ofUnsignedShort(BLS_PRIME_3),
```
Bytes.ofUnsignedShort is always 2 bytes, prefer Bytes.minimal() that trimm the leading 0s

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/bls/BlsOperation.java:
```diff
@@ -0,0 +1,1094 @@
+    isInfinity = repeat(false, nRows);
+    nontrivialPairOfPointsBit = repeat(false, nRows);
+
+    wcpFlag = repeat(false, nRows);
```
you would avoid all of this ugly stuff if you just store a list of wcp call, as told https://github.com/Consensys/linea-tracer/pull/2092/files#r2158544957

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/bls/BlsOperation.java:
```diff
@@ -0,0 +1,1094 @@
+    final boolean mextBit = internalChecksPassed && !successBit;
+
+    for (int j = 0; j <= CT_MAX_POINT_EVALUATION; j++) {
+      this.mintBit.set(j, !internalChecksPassed);
```
not sure, but can't it be defcomputedcolumn by corset directly ?

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/bls/BlsOperation.java:
```diff
@@ -0,0 +1,1094 @@
+    }
+  }
+
+  void trace(Trace.Bls trace, final int stamp, final long previousId) {
```
I think it would be easier to read if you have an abstract BlsOperation that get implemented by each different precompile. Each class would have only it's own method instead of having a huge switch case for everything

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/bls/BlsOperation.java:
```diff
@@ -0,0 +1,1094 @@
+    boolean nontrivialPairOfPointsAcc = false;
+
+    for (int i = 0; i < nRows; i++) {
+      boolean isData = i < nRowsData;
```
why not having two tracing method, one for the data, one for the result ? It would be way easier to read, no more need of isData ? at each row +. resetting ct would be trivial.

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/bls/BlsOperation.java:
```diff
@@ -0,0 +1,1094 @@
+          .limb(isData || returnDataIsNonEmpty ? limb.slice(i, LLARGE) : ZERO)
+          .successBit(successBit)
+          .ct(ct)
+          .ctMax(ctMax)
```
can't CtMAx be a defComputedColumn ?

lorenzogentile404 replied:
Likely yes. I'll check it out.

letypequividelespoubelles reviewed (commented):
Some refcato would be nice in BLS to have smaller and easily readable class + less switch.

+ some comments

amkCha committed:
feat: re-enable cancun unit tests refering to KZG precomp

lorenzogentile404 committed:
removed imeplementation of bls elliptic curve arithmetic implementation as we use the one from Besu

lorenzogentile404 committed:
Merge branch 'arith-dev' into eip-2537-precompile-for-bls12-381-curve-operations

lorenzogentile404 committed:
Merge branch 'arith-dev' into eip-2537-precompile-for-bls12-381-curve-operations

lorenzogentile404 committed:
partially debugged point evaluation

lorenzogentile404 committed:
correctly initialized effective calls to bls data

lorenzogentile404 committed:
updated constraints

lorenzogentile404 committed:
solved bug related to isFirstInput

lorenzogentile404 committed:
Merge branch 'arith-dev' into eip-2537-precompile-for-bls12-381-curve-operations

lorenzogentile404 committed:
fixed bug related to stamp constancy of mint and mext

lorenzogentile404 committed:
added header

lorenzogentile404 committed:
updated constraints

lorenzogentile404 committed:
updated constraints

lorenzogentile404 committed:
fixed bug related to mext bit and mext acc

lorenzogentile404 committed:
updated constraints

lorenzogentile404 committed:
enabled more test instanced for point evaluation

lorenzogentile404 committed:
restricted bls tests to point evaluation

lorenzogentile404 committed:
fixed call to point evaluation precompile due to missing initialization

lorenzogentile404 committed:
fixed minor issues

lorenzogentile404 committed:
Merge branch 'arith-dev' into eip-2537-precompile-for-bls12-381-curve-operations

lorenzogentile404 committed:
solved bug related to tracing of limb and reached black belt level in debugging

lorenzogentile404 committed:
fixed bug related to the computation of wtr

lorenzogentile404 committed:
fixed bug related to tracing of precompile scenario

lorenzogentile404 committed:
updated constraints

lorenzogentile404 committed:
added point evaluation tests

lorenzogentile404 committed:
Merge branch 'arith-dev' into eip-2537-precompile-for-bls12-381-curve-operations

lorenzogentile404 committed:
Merge branch 'arith-dev' into eip-2537-precompile-for-bls12-381-curve-operations

lorenzogentile404 committed:
add SixtyThreeSixtyFourthsTests tests for POINT_EVALUATION

lorenzogentile404 committed:
Merge branch 'arith-dev' into eip-2537-precompile-for-bls12-381-curve-operations

lorenzogentile404 committed:
switched to Cancun

lorenzogentile404 committed:
added temporary test

lorenzogentile404 committed:
Merge branch 'arith-dev' into eip-2537-precompile-for-bls12-381-curve-operations

lorenzogentile404 committed:
reverted constraints update

lorenzogentile404 committed:
updated constraints and cleaned code

lorenzogentile404 committed:
spotless

lorenzogentile404 committed:
removed useless check

lorenzogentile404 committed:
updated constraints

lorenzogentile404 committed:
clened code

lorenzogentile404 committed:
checks on small and large points take the whole point as input bytes

lorenzogentile404 committed:
clened code

lorenzogentile404 committed:
not splitting anymore inputs in blsdataoperation

lorenzogentile404 committed:
Merge branch 'arith-dev' into eip-2537-precompile-for-bls12-381-curve-operations

lorenzogentile404 committed:
Merge branch 'arith-dev' into eip-2537-precompile-for-bls12-381-curve-operations

lorenzogentile404 committed:
updated constraints

lorenzogentile404 committed:
updated constraints

lorenzogentile404 committed:
updated constraints

lorenzogentile404 committed:
debugged replay tests

lorenzogentile404 committed:
fixed point evalution tests
